File: src/openprecedent/schemas/event.py
from datetime import datetime
from enum import StrEnum

from pydantic import BaseModel, ConfigDict, Field


class EventType(StrEnum):
    CASE_STARTED = "case.started"
    CHECKPOINT_SAVED = "checkpoint.saved"
    MESSAGE_USER = "message.user"
    MESSAGE_AGENT = "message.agent"
    MODEL_INVOKED = "model.invoked"
    MODEL_COMPLETED = "model.completed"
    TOOL_CALLED = "tool.called"
    TOOL_COMPLETED = "tool.completed"
    COMMAND_STARTED = "command.started"
    COMMAND_COMPLETED = "command.completed"
    FILE_READ = "file.read"
    FILE_WRITE = "file.write"
    USER_CONFIRMED = "user.confirmed"
    CASE_COMPLETED = "case.completed"
    CASE_FAILED = "case.failed"


class EventActor(StrEnum):
    USER = "user"
    AGENT = "agent"
    SYSTEM = "system"
    TOOL = "tool"


class Event(BaseModel):
    model_config = ConfigDict(extra="forbid")

    event_id: str
    case_id: str
    event_type: EventType
    actor: EventActor
    timestamp: datetime
    sequence_no: int
    parent_event_id: str | None = None
    payload: dict[str, object] = Field(default_factory=dict)

File: rust/openprecedent-contracts/src/event.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use strum::{Display, EnumString};

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize, Display, EnumString)]
pub enum EventType {
    #[serde(rename = "case.started")]
    #[strum(serialize = "case.started")]
    CaseStarted,
    #[serde(rename = "checkpoint.saved")]
    #[strum(serialize = "checkpoint.saved")]
    CheckpointSaved,
    #[serde(rename = "message.user")]
    #[strum(serialize = "message.user")]
    MessageUser,
    #[serde(rename = "message.agent")]
    #[strum(serialize = "message.agent")]
    MessageAgent,
    #[serde(rename = "model.invoked")]
    #[strum(serialize = "model.invoked")]
    ModelInvoked,
    #[serde(rename = "model.completed")]
    #[strum(serialize = "model.completed")]
    ModelCompleted,
    #[serde(rename = "tool.called")]
    #[strum(serialize = "tool.called")]
    ToolCalled,
    #[serde(rename = "tool.completed")]
    #[strum(serialize = "tool.completed")]
    ToolCompleted,
    #[serde(rename = "command.started")]
    #[strum(serialize = "command.started")]
    CommandStarted,
    #[serde(rename = "command.completed")]
    #[strum(serialize = "command.completed")]
    CommandCompleted,
    #[serde(rename = "file.read")]
    #[strum(serialize = "file.read")]
    FileRead,
    #[serde(rename = "file.write")]
    #[strum(serialize = "file.write")]
    FileWrite,
    #[serde(rename = "user.confirmed")]
    #[strum(serialize = "user.confirmed")]
    UserConfirmed,
    #[serde(rename = "case.completed")]
    #[strum(serialize = "case.completed")]
    CaseCompleted,
    #[serde(rename = "case.failed")]
    #[strum(serialize = "case.failed")]
    CaseFailed,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize, Display, EnumString)]
#[serde(rename_all = "snake_case")]
#[strum(serialize_all = "snake_case")]
pub enum EventActor {
    User,
    Agent,
    System,
    Tool,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct Event {
    pub event_id: String,
    pub case_id: String,
    pub event_type: EventType,
    pub actor: EventActor,
    pub timestamp: DateTime<Utc>,
    pub sequence_no: i64,
    pub parent_event_id: Option<String>,
    pub payload: Value,
}
